src/plugins/grpc/client.js

```diff
@@ -0,0 +1,143 @@
+'use strict'
+
+const FORMAT_TEXT_MAP = require('opentracing').FORMAT_TEXT_MAP
+const log = require('../../log')
+const pick = require('lodash.pick')
+const Tags = require('opentracing').Tags
+
+function getFilter (config, filter) {
+  if (typeof config[filter] === 'function') {
+    return config[filter]
+  }
+
+  if (config[filter] instanceof Array) {
+    return element => pick(element, config[filter])
+  }
+
+  if (config.hasOwnProperty(filter)) {
+    log.error(`Expected '${filter}' to be an array or function.`)
+  }
+
+  return null
+}
+
+function getMethodType (definition) {
+  if (definition.requestStream) {
+    if (definition.responseStream) {
+      return 'bidi'
+    }
+
+    return 'client_stream'
+  }
+
+  if (definition.responseStream) {
+    return 'server_stream'
+  }
+
+  return 'unary'
+}
+
+function patch (grpc, tracer, config) {
+  if (config.client === false) return
+
+  config = config.client || config
+
+  const configFields = getFilter(config, 'fields')
+  const configMetadata = getFilter(config, 'metadata')
+
+  function datadogInterceptor (options, nextCall) {
+    const path = options.method_definition.path
+    const methodType = getMethodType(options.method_definition)
+    const scope = tracer.scopeManager().active()
+    const span = tracer.startSpan('grpc.request', {
+      childOf: scope && scope.span(),
+      tags: {
+        [Tags.SPAN_KIND]: Tags.SPAN_KIND_RPC_CLIENT,
+        'grpc.method.name': path,
+        'grpc.method.type': methodType,
+        'resource.name': path.split('/').slice(-1),
+        'service.name': config.service || `${tracer._service}-grpc-client`
+      }
+    })
+
+    return new grpc.InterceptingCall(nextCall(options), {
+      sendMessage: (message, next) => {
+        if (configFields) {
+          const values = configFields(message)
+
+          for (const key in values) {
+            span.setTag(`grpc.request.message.fields.${key}`, values[key])
+          }
+        }
+
+        next(message)
+      },
+
+      cancel: () => {
+        span.setTag('cancelled', true)
```

Comment:
This should be scoped like the other tags. Also, would it be better to have some kind of "state" with different values instead of multiple tags per state?

Reply:
I managed to get a nice textual status of each request for almost every case, except _unary_ calls on the server plugin. :(

Spent a lot longer on it than I care to admit 🙈but can't really find a solution the way that grpc is set up. Hopefully when we have server interceptors it'll improve.

Reply:
A few things about the new implementation:

• Would it be possible to use the keys of the constants instead of strings?
• Can something like [this](https://github.com/grpc/grpc-node/blob/633736225ac4f8ce05bd246ffe576616e9290615/packages/grpc-native-core/src/server.js#L44-L70) be used to categorize errors?
• Since the code can be set by the user to a value outside of the known codes, I would say it would be best to have the code as a number instead of a string. We could always include the string as well for known codes in another tag.

Reply:
> Can something like this be used to categorize errors?

Sorry, not sure what you mean here. 

> Since the code can be set by the user to a value outside of the known codes, I would say it would be best to have the code as a number instead of a string.

Makes sense. I've updated the PR.

> We could always include the string as well for known codes in another tag.

Agreed. I've left it out for now but it should be trivial to add if needed.

Reply:
> Sorry, not sure what you mean here.

I think I misunderstood what you meant by `I managed to get a nice textual status of each request for almost every case, except unary calls on the server plugin.`

Does this mean that for unary calls it's not possible to get the status code at all? How useful would the status code be for these calls?

Reply:
> Does this mean that for unary calls it's not possible to get the status code at all?

It's possible on the _client_ plugin (since the client interceptor API exposes the status code) but not on the _server_ plugin.

> How useful would the status code be for these calls?

We know whether the unary call returned an error, so:

1. If it didn't return an error (was successful) we can pretty much guarantee it was OK (code `0`).
2. If it **did** return an error, I was defaulting to `UNKNOWN` but it's not technically correct. It could also be INVALID_ARGUMENT, NOT_FOUND, etc.

Reply:
My understanding is that the status code comes from the error object for errors, so it may be available at that point. Not 100% sure though.

Reply:
That's technically true but the function I'm wrapping doesn't give me a chance to grab the error object with the code. I'd need to be able to grab the `err` variable from [here](https://github.com/grpc/grpc-node/blob/master/packages/grpc-native-core/src/server.js#L575-L590), but I'm wrapping the `func()` call which doesn't get access to it. 

At the point where my function is called, `emitter.error` doesn't have a code property.

Reply:
Would wrapping `call.startBatch` as well work in this case?

Reply:
Looks like it would be difficult to do that properly. In that case I'd probably still check for `error.code` since it could have been set elsewhere, and if there is no `error.code` then skip setting the tag and rely only on the information provided by the error tags.

Reply:
I spent a lot of time trying to wrap `call.startBatch` early on (as well as the `handleUnary`, `handleServerStreaming`, etc calls but with little success :(

While grpc is indeed attempting to copy the `code` from the error object ([if it exists](https://github.com/grpc/grpc-node/blob/master/packages/grpc-native-core/src/server.js#L53)) I cannot reproduce this behaviour in a test.

From what I can gather it's meant to be provided by the caller so I guess it makes sense to honour it if it's there.

PR updated.

Reply:
I prefer to not set the tag instead of using `UNKNOWN` when `error.code` is not set since it could be confusing if there is actually a status code but it's different.